src/Sarif.Multitool/Rules/SARIF2004.OptimizeFileSize.cs

```diff
@@ -41,20 +41,94 @@ public class OptimizeFileSize : SarifValidationSkimmerBase
         /// the 'artifacts' array should be retained even if it contains only location information.
+        /// 
+        /// In addition to the avoiding unnecessary arrays, there are other ways to optimize the
+        /// size of SARIF log files.
+        /// 
+        /// Prefer the result object properties 'ruleId' and 'ruleIndex' to the nested object-valued
+        /// property 'result.rule', unless the rule comes from a tool component other than the driver
+        /// (in which case only 'result.rule' can accurately point to the metadata for the rule).
+        /// The 'ruleId' and 'ruleIndex' properties are shorter and just as clear.
+        /// 
+        /// Do not specify the result object's 'analysisTarget' property unless it differs from the
+        /// result location. The canonical scenario for using 'result.analysisTarget' is a C/C++ language
+        /// analyzer that is instructed to analyze example.c, and detects a result in the included file
+        /// example.h. In this case, 'analysisTarget' is example.c, and the result location is in example.h.
         /// </summary>
         public override MultiformatMessageString FullDescription => new MultiformatMessageString { Text = RuleResources.SARIF2004_OptimizeFileSize_FullDescription_Text };
 
         protected override IEnumerable<string> MessageResourceNames => new string[] {
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeAnalysisTarget_Text),
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_AvoidDuplicativeResultRuleInformation_Text),
             nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateLocationOnlyArtifacts_Text),
-            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text)
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_EliminateIdOnlyRules_Text),
+            nameof(RuleResources.SARIF2004_OptimizeFileSize_Warning_PreferRuleId_Text)
         };
 
         public override FailureLevel DefaultLevel => FailureLevel.Warning;
 
+        private string driverGuid;
+
         protected override void Analyze(Run run, string runPointer)
         {
             AnalyzeLocationOnlyArtifacts(run, runPointer);
             AnalyzeIdOnlyRules(run, runPointer);
+
+            this.driverGuid = run.Tool.Driver?.Guid;
+        }
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            ReportUnnecessaryAnalysisTarget(result, resultPointer);
+            ReportRuleDuplication(result, resultPointer);
+        }
+
+        private void ReportRuleDuplication(Result result, string resultPointer)
+        {
+            if (result.Rule != null)
+            {
+                if (result.Rule.ToolComponent != null)
+                {
+                    if (result.Rule.ToolComponent.RefersToDriver(this.driverGuid))
```

Comment:
You can shorten this:
```
if (result.Rule.ToolComponent?.RefersToDriver(this.driverGuid) == true)
```

Note that you need the explicit comparison `== true` because the null-coalescing operator `?.` causes the left-hand side to have the type `Nullable<bool>` -- and you can't do an implicit logical test on a `Nullable<bool>`.